app/src/org/commcare/android/database/HybridFileBackedSqlHelpers.java

```diff
@@ -0,0 +1,108 @@
+ * store serialized object payloads.
+ *
+ * @author Phillip Mates (pmates@dimagi.com).
+ */
+public class HybridFileBackedSqlHelpers {
+
+    protected static Pair<String, byte[]> getEntryFilenameAndKey(AndroidDbHelper helper,
+                                                                 String table,
+                                                                 int id) {
+        Cursor c;
+        try {
+            c = helper.getHandle().query(table, HybridFileBackedSqlStorage.dataColumns,
+                    DatabaseHelper.ID_COL + "=?",
+                    new String[]{String.valueOf(id)}, null, null, null);
+        } catch (SessionUnavailableException e) {
+            throw new UserStorageClosedException(e.getMessage());
+        }
+
+        try {
+            c.moveToFirst();
+            return new Pair<>(c.getString(c.getColumnIndexOrThrow(DatabaseHelper.FILE_COL)),
+                    c.getBlob(c.getColumnIndexOrThrow(DatabaseHelper.AES_COL)));
+        } finally {
+            if (c != null) {
+                c.close();
+            }
+        }
+    }
+
+    protected static String getEntryFilename(AndroidDbHelper helper,
+                                             String table, int id) {
+        Cursor c;
+        SQLiteDatabase db;
+        try {
+            db = helper.getHandle();
+        } catch (SessionUnavailableException e) {
+            throw new UserStorageClosedException(e.getMessage());
+        }
+
+        String[] columns = new String[]{DatabaseHelper.FILE_COL};
+        c = db.query(table, columns, DatabaseHelper.ID_COL + "=?",
+                new String[]{String.valueOf(id)}, null, null, null);
+
+        try {
+            c.moveToFirst();
+            return c.getString(c.getColumnIndexOrThrow(DatabaseHelper.FILE_COL));
+        } finally {
+            if (c != null) {
+                c.close();
+            }
+        }
+    }
+
+    protected static void removeFiles(List<Integer> idsBeingRemoved,
+                                      AndroidDbHelper helper,
+                                      String table) {
+        // delete files storing data for entries being removed
+        for (Integer id : idsBeingRemoved) {
+            String filename = HybridFileBackedSqlHelpers.getEntryFilename(helper, table, id);
+            if (filename != null) {
+                File datafile = new File(filename);
+                datafile.delete();
+            }
+        }
+    }
+
+    protected static File newFileForEntry(File dbDir) throws IOException {
+        File newFile = getUniqueFilename(dbDir);
+
+        if (!newFile.createNewFile()) {
+            throw new RuntimeException("Trying to create a new file using existing filename; " +
+                    "Shouldn't be possible since we already checked for uniqueness");
+        }
+
+        return newFile;
+    }
+
+    private static File getUniqueFilename(File dbDir) {
+        String uniqueSuffix = "_" + (Math.random() * 5);
```

Comment:
Can we just use GUID's for these? seems more reliable/consistent than making up a structure.